File: ch5/src/py_demo_tf/py_demo_tf/dynamic_tf_broadcaster.py
import rclpy
from rclpy.node import Node 
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped
from tf_transformations import quaternion_from_euler

class DynamicTFBroadcaster(Node):
    def __init__(self):
        super().__init__('dynamic_tf_broadcaster')
        self.tf_broadcaster_ = TransformBroadcaster(self)
        self.timer_ = self.create_timer(1, self.publish_transform)
        
    def publish_transform(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "camera_link"
        transform.child_frame_id = "bottle_link"
        
        transform.transform.translation.x = 0.2
        transform.transform.translation.y = 0.0
        transform.transform.translation.z = 0.5
        
        rotation_quaternion = quaternion_from_euler(0, 0, 0)
        transform.transform.rotation.x = rotation_quaternion[0]
        transform.transform.rotation.y = rotation_quaternion[1]
        transform.transform.rotation.z = rotation_quaternion[2]
        transform.transform.rotation.w = rotation_quaternion[3]
        
        self.tf_broadcaster_.sendTransform(transform)
        

def main():
    rclpy.init()
    tf_node = DynamicTFBroadcaster()
    rclpy.spin(tf_node)
    rclpy.shutdown()
        

File: ch5/src/cpp_demo_tf/src/dynamic_tf_broadcaster.cpp
#include <memory>
#include <chrono>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/transform_broadcaster.h"

using namespace std::chrono_literals;

class DynamicTFBroadcaster: public rclcpp::Node {
private:
    std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    rclcpp::TimerBase::SharedPtr timer_;

public:
    DynamicTFBroadcaster(): Node("dynamic_tf_broadcaster") {
        this->tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        this->timer_ = this->create_wall_timer(10ms, std::bind(&DynamicTFBroadcaster::publish_transform, this));
    }

    void publish_transform() {
        geometry_msgs::msg::TransformStamped transform;
        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "base_link";
        
        transform.transform.translation.x = 2.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;

        tf2::Quaternion quaternion;
        quaternion.setRPY(0, 0, 30 * M_PI / 180);
        transform.transform.rotation = tf2::toMsg(quaternion);

        this->tf_broadcaster_->sendTransform(transform);
    }
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DynamicTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}